plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3Service.java

```diff
@@ -238,104 +265,141 @@ protected PasswordAuthentication getPasswordAuthentication() {
-        clientConfiguration.setUseThrottleRetries(clientSettings.throttleRetries);
-        clientConfiguration.setSocketTimeout(clientSettings.readTimeoutMillis);
+        Protocol proxyProtocol = clientSettings.proxySettings.getType() == ProxySettings.ProxyType.DIRECT
+            ? Protocol.HTTP
+            : clientSettings.proxySettings.getType().toProtocol();
+
+        try {
+            proxyConfiguration = proxyConfiguration.endpoint(
+                new URI(
+                    proxyProtocol.toString(),
+                    null,
+                    clientSettings.proxySettings.getHost(),
+                    clientSettings.proxySettings.getPort(),
+                    null,
+                    null,
+                    null
+                )
+            );
+        } catch (URISyntaxException e) {
+            throw new RuntimeException("Invalid proxy URL", e);
+        }
+
+        proxyConfiguration = proxyConfiguration.username(clientSettings.proxySettings.getUsername());
+        proxyConfiguration = proxyConfiguration.password(clientSettings.proxySettings.getPassword());
+
+        return proxyConfiguration.build();
+    }
 
-        return clientConfiguration;
+    static ClientOverrideConfiguration buildOverrideConfiguration(final S3ClientSettings clientSettings) {
+        ClientOverrideConfiguration.Builder clientOverrideConfiguration = ClientOverrideConfiguration.builder();
+        if (Strings.hasLength(clientSettings.signerOverride)) {
+            clientOverrideConfiguration = clientOverrideConfiguration.putAdvancedOption(
+                SdkAdvancedClientOption.SIGNER,
+                AwsRequestSigner.fromSignerName(clientSettings.signerOverride).getSigner()
+            );
+        }
+        RetryPolicy.Builder retryPolicy = RetryPolicy.builder().numRetries(clientSettings.maxRetries).retryCapacityCondition(null);
+        if (!clientSettings.throttleRetries) {
+            retryPolicy.throttlingBackoffStrategy(BackoffStrategy.none());
+        }
+        return clientOverrideConfiguration.retryPolicy(retryPolicy.build()).build();
     }
 
     private static SSLConnectionSocketFactory createSocksSslConnectionSocketFactory(final InetSocketAddress address) {
         // This part was taken from AWS settings
-        final SSLContext sslCtx = SdkSSLContext.getPreferredSSLContext(
-            new SystemPropertyTlsKeyManagersProvider().getKeyManagers(),
-            new SecureRandom()
-        );
-        return new SdkTLSSocketFactory(sslCtx, new DefaultHostnameVerifier()) {
-            @Override
-            public Socket createSocket(final HttpContext ctx) throws IOException {
-                return new Socket(new Proxy(Proxy.Type.SOCKS, address));
-            }
-        };
+        try {
+            final SSLContext sslCtx = SSLContext.getDefault();
+            sslCtx.init(SystemPropertyTlsKeyManagersProvider.create().keyManagers(), null, new SecureRandom());
```

Comment:
It seems like we are missing `SystemPropertyTlsTrustManagersProvider` (no in SDK), should we introduce one?

Reply:
Currently, we use [`SdkSSLContext.getPreferredSSLContext`](https://github.com/opensearch-project/OpenSearch/blob/adf7e2cdfbc13013448bc07ddac0fe09c07fd7e5/plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3Service.java#L268) from the v1 SDK to get an `SSLContext`, which [internally sets the trust managers parameter to `null`](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-core/src/main/java/com/amazonaws/internal/SdkSSLContext.java#L38) while initializing it. The behaviour here is exactly the same. Do you foresee a need to add a `SystemPropertyTlsTrustManagersProvider`?

While going through this I realized that the [v1 SDK was explicitly specifying `TLS` as the protocol](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-core/src/main/java/com/amazonaws/internal/SdkSSLContext.java#L36) while fetching the SSLContext, while we are currently doing a [`SSLContext.getDefault()`](https://github.com/opensearch-project/OpenSearch/pull/7372/files?diff=split&w=1#diff-b8fbfee9d71f7f2caf16514727692e968b4638afcafce12b4698e2046e6cb673R330) call, which is a deviation, will fix this.

Reply:
This is a good point, I think just `null` there makes it explicit that trust managers are not provided, but I think we could keep the behavior as it was before, with `null`, thanks for clarifying.